mllib/src/main/scala/org/apache/spark/ml/feature/Imputer.scala

```diff
@@ -0,0 +1,231 @@
+  extends Estimator[ImputerModel] with ImputerParams with DefaultParamsWritable {
+
+  @Since("2.0.0")
+  def this() = this(Identifiable.randomUID("imputer"))
+
+  /** @group setParam */
+  def setInputCol(value: String): this.type = set(inputCol, value)
+
+  /** @group setParam */
+  def setOutputCol(value: String): this.type = set(outputCol, value)
+
+  /**
+   * Imputation strategy. Available options are ["mean", "median" and "mode"].
+   * @group setParam
+   */
+  def setStrategy(value: String): this.type = set(strategy, value)
+
+  /** @group setParam */
+  def setMissingValue(value: Double): this.type = set(missingValue, value)
+
+  setDefault(strategy -> "mean", missingValue -> Double.NaN)
+
+  override def fit(dataset: Dataset[_]): ImputerModel = {
+    transformSchema(dataset.schema, logging = true)
+    val ic = col($(inputCol))
+    val filtered = dataset.select(ic.cast(DoubleType))
+      .filter(ic.isNotNull && ic =!= $(missingValue))
+    val surrogate = $(strategy) match {
+      case "mean" => filtered.filter(!ic.isNaN).select(avg($(inputCol))).first().getDouble(0)
+      case "median" => filtered.stat.approxQuantile($(inputCol), Array(0.5), 0.001)(0)
+      case "mode" => filtered.rdd.map(r => r.getDouble(0)).map(d => (d, 1)).reduceByKey(_ + _)
+        .sortBy(-_._2).first()._1
+    }
+    copyValues(new ImputerModel(uid, surrogate).setParent(this))
+  }
+
+  override def transformSchema(schema: StructType): StructType = {
+    validateAndTransformSchema(schema)
+  }
+
+  override def copy(extra: ParamMap): Imputer = {
+    val copied = new Imputer(uid)
+    copyValues(copied, extra)
+  }
+}
+
+@Since("2.0.0")
+object Imputer extends DefaultParamsReadable[Imputer] {
+
+  /** Set of strategy names that Imputer currently supports. */
+  private[ml] val supportedStrategyNames = Set("mean", "median", "mode")
+
+  @Since("2.0.0")
+  override def load(path: String): Imputer = super.load(path)
+}
+
+/**
+ * :: Experimental ::
+ * Model fitted by [[Imputer]].
+ *
+ * @param surrogate statistics value for each feature during fitting
+ */
+@Experimental
+class ImputerModel private[ml](
+    override val uid: String,
+    val surrogate: Double)
+  extends Model[ImputerModel] with ImputerParams with MLWritable {
+
+  import ImputerModel._
+
+  /** @group setParam */
+  def setInputCol(value: String): this.type = set(inputCol, value)
+
+  /** @group setParam */
+  def setOutputCol(value: String): this.type = set(outputCol, value)
+
+  override def transform(dataset: Dataset[_]): DataFrame = {
+    transformSchema(dataset.schema, logging = true)
+    val inputType = dataset.select($(inputCol)).schema.fields(0).dataType
+    inputType match {
+      case _: NumericType =>
+        val ic = col($(inputCol)).cast(DoubleType)
+        dataset.withColumn($(outputCol), when(ic.isNull, surrogate)
+          .when(ic === $(missingValue), surrogate)
+          .otherwise(ic)
+          .cast(inputType))
```

Comment:
Just catching up now... I like the idea of maintaining the input type.  I'm imagining using an Imputer to fill in continuous features with the mean and categoricals with the mode.  Later on, we could even check to see if a column is categorical (in the metadata) and throw an exception for mean.

I'd prefer your option (b) to be safe.